src/lib/datepicker/date-selection.ts

```diff
@@ -0,0 +1,137 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {Subject} from 'rxjs/Subject';
+import {DateAdapter} from '@angular/material/core';
+import {EventEmitter} from '@angular/core';
+
+/**
+ * Special interface for dates interval.
+ */
+export interface MatDatepickerRange<D> {
```

Comment:
Lets make this an actual class (`DateRange<D>`) so we can do `value instanceof DateRange`. This is safer than checking for `begin` and `end` properties because `D` is generic and could theoretically have those same properties.